Limit memory used while fetching many partitions

dlex opened the pull request:
A semaphore is added to `kafka::server` to control the memory used for the buffers in the fetch path. The semaphore is set to a configurable fraction of kafka shard memory (default 0.5). As the fetch work is delegated to different shards, the first partition requested from a shard is always served, other partitions may be skipped if the memory semaphore on the shard does not have enough units available. Memory reservation is done based on the estimation of 1 MiB per batch (default, configurable), which is adjusted as soon as the real used size is known. The semaphore is local to the partition home shard.

The overall kafka server memory semaphore is also used the same way, so sizes of fetch responses are also limited from the overall kafka server perspective as well. It is also the partition home shard's semaphore, not the one associated with the connection that has received the fetch request. This is done because the memory for the buffers is acquired on the partition home shard.

Fetch request memory estimator is not changed, because it only works with the memory semaphore associated with the connection shard, and at the time of the estimation we don't even know what shards will be affected.

Neither of the semaphores is waited upon, the amounts only get consumed from there. Both semaphores can go negative as the result. So, when overused:
* fetch memory semaphore does not delay anything, but `read_from_ntp`s will only perform the minimal read then (one batch per shard's part of the plan)
* kafka server memory semaphore will do the same, and additionally it will delay incoming requests served by the shard.

Actually, fetch memory semaphore is never waited upon and could have been an integer instead. The only reason it is a semaphore is a convenience of `semaphore_units`.

As a side effect, this PR adds support for floating point bounded properties.

The differences from the reverted #10371 are:
* In 10371, both semaphores decremented before reading from partition were the ones associated with the connection shard; now they both are the ones associated with the partition home shard. Besides the fact that was a logic error, in 10371 that also caused the kafka server memory semaphore to be used in both connection shard (for request memory estimation), and in other (partition) shards. Allegedly due to promises in the semaphore waitlist created and signalled in different shard, there was memory corruption that was manifested in an unrelated location.
* In 10371 there was an attempt made to estimate request memory with a custom `fetch_memory_estimator`, not done in this PR because memory estimator only affects connection shard memory semaphore, while the idea of the PR is to account memory against the memory semaphores of the shards where the memory is getting allocated.

Fixes #3409.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v23.1.x
- [x] v22.3.x
- [ ] v22.2.x

## Release Notes

### Improvements

* Memory control is improved in Fetch request handler, covering the cases when a client tries to fetch too many partitions lead by the same shard. Some of the request partitions will not be fetched if the broker does not have enough memory for that, or if that would violate constraints set by `kafka_max_bytes_per_fetch` and `fetch_max_bytes`.

* If `kafka_max_bytes_per_fetch` is not configured properly, redpanda is now more robust against huge Fetch requests by controlling memory consumption dynamically.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

dlex commented:
CI failure in https://buildkite.com/redpanda/redpanda/builds/29645#0188471d-0cf2-4691-ad2e-4722e90f3f7f:
* existing: #10497

dlex commented:
/ci-repeat 3

dlex commented:
CI failures in https://buildkite.com/redpanda/redpanda/builds/29714#01884a44-7df4-4a4f-8a28-16a952c0ee31:
* existing: #8698

in https://buildkite.com/redpanda/redpanda/builds/29714#01884a37-0b3d-4a6c-9a29-322f4cb9d22d
* existing: #10865

dlex commented:
Force push: rebased upstream

dlex commented:
Force push: linter errors

jcsp commented:
@dlex could you summarize the changes between the previous reverted change and this one?  I'm curious what the issue was.

dlex commented:
> @dlex could you summarize the changes between the previous reverted change and this one? I'm curious what the issue was.

@jcsp I've added the summary to the PR description.

dlex commented:
CI failures in https://buildkite.com/redpanda/redpanda/builds/29944#0188567a-a1c7-4d0a-9cd8-402aa44074ee:
* existing: #10502

dlex commented:
CI failure in https://buildkite.com/redpanda/redpanda/builds/29944#01885699-716c-4fc4-bb28-fd604f647434:
* existing: #10868

dlex commented:
Force push: rebased upstream

dlex commented:
Force push: some comments revisited

dlex commented:
Force push: test improvements
* raised the upper limit for `memory_share_for_fetch` to 0.95; it is likely that it failed for tha values above 0.8 befere because of the logic error
* unit test: now memory units returned from both memory semaphores are the same

dlex commented:
CI failure in https://buildkite.com/redpanda/redpanda/builds/30076#018863d1-7035-46dd-b6c8-1a0de1f45a6d:
* existing #10865

dlex commented:
CI failures in https://buildkite.com/redpanda/redpanda/builds/30076#018863d1-7037-485f-a7e0-202b0df3ddae:
* existing: #10502
* existing: #11044

dlex commented:
Force push: maximum value for `memory_share_for_fetch` is set to 0.9

dlex commented:
Force push: maximum value for `memory_share_for_fetch` is set to 0.8

dotnwat commented:
> @dlex - I couldn't quite determine how this works from the description in the PR. Can you lengthen it a bit?

@travisdowns it looks like the cover letter was updated in case you missed the notification.

dotnwat commented:
/dt-repeat 10
skip-units

dotnwat commented:
/ci-repeat 10

dlex commented:
CI failures in debug CI of https://buildkite.com/redpanda/redpanda/builds/30115#_:
* existing #10867 in https://buildkite.com/redpanda/redpanda/builds/30115#01886a62-aceb-486b-84cd-2c58eda62112

dlex commented:
Force push: rebased upstream

dlex commented:
/ci-repeat 3
dt-repeat 30
skip-unit
tests/rptest/tests/follower_fetching_test.py::FollowerFetchingTest.test_basic_follower_fetching

dotnwat commented:
>As the fetch work is delegated to different shards, the first partition requested from a shard is always served, other partitions may be skipped

what does it mean to be skipped?

dotnwat commented:
>Neither of the semaphores is waited upon, the amounts only get consumed from there. Both semaphores can go negative as the result. So, when overused:
fetch memory semaphore does not delay anything, but read_from_ntps will only perform the minimal read then (one batch per shard's part of the plan)

do i understand correctly that basically when a semaphore is exhausted, fetches will be limited to returning a single batch (total or per partition or per shard)?

dlex commented:
> what does it mean to be skipped?

It means to include the partition record into the response with valid offsets but with no messages fetched: https://github.com/redpanda-data/redpanda/blob/dev/src/v/kafka/server/handlers/fetch.cc#L82-L86

dlex commented:
> do i understand correctly that basically when a semaphore is exhausted, fetches will be limited to returning a single batch (total or per partition or per shard)?

right, fetches will return a single batch per each shard that perticipates in the fetch and has either of the semahores exhausted.

dotnwat commented on src/v/kafka/server/request_context.h:
```diff
@@ -246,6 +246,16 @@ class request_context {
         return _conn->server().get_replica_selector();
     }
 
+    /// \ref kafka::server memory semaphore for the local shard
+    ssx::semaphore& memory_sem_local() const noexcept {
+        return _conn->server().container().local().memory();
+    }
+
+    /// \ref kafka::server fetch memory semaphore for the local shard
+    ssx::semaphore& memory_fetch_sem_local() const noexcept {
+        return _conn->server().container().local().memory_fetch_sem();
+    }
```
can you add some emphasis to the comments that the local shard is for the _caller_? we don't really have any sort of formal rules around it, but generally working directly with the sharded type in these cases is nice because it forces the caller to explicitly write `.local()`. So alternatively another way to improve this interface other than a comment woudl be to do something like

```
ss::sharded<server>& server() return { _conn->server().container(); }
// at the caller
server().local().memory_fetch_sem();
```

dlex replied:
I like the way you have suggested. I did not do that because I thought the design was to expose only an explit API from `request_context`, but if it's not, then `ss::sharded<server>& server()` is a much better way

dotnwat replied:
it's all good. i think the design so far just didn't have a use case like this.

dotnwat commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -443,7 +564,9 @@ handle_shard_fetch(ss::shard_id shard, op_context& octx, shard_fetch fetch) {
          &replica_selector = octx.rctx.replica_selector(),
-         bytes_left = octx.bytes_left](
+         bytes_left = octx.bytes_left,
+         &memory_sem = octx.rctx.memory_sem_local(),
+         &memory_fetch_sem = octx.rctx.memory_fetch_sem_local()](
```
this looks to me like a reference to the semaphores on the calling shard is going to be passed to the destination shard, operating under the assumption that the lambda captures is created on the originating core (which is how i'm imagining that it works). can you confirm that's the case or not?

dotnwat commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -343,6 +343,8 @@ static ss::future<read_result> do_read_from_ntp(
     co_return result;
 }
 
+namespace testing {
```
+1

dotnwat reviewed (commented):
looks pretty good.

i think there are still two unsafe cross-core memory references. one that i noticed existed before your PR, and then a new one. i left a comment about both. i could be mistaken about how things are working, so it'd be great if you could confirm/reject the idea that the unsafe access exists.

hoping too that you could characterize the testing that is in place for this feature. i see there are both UI and DT tests (yay), but it'd be nice to understand how aggressive/complete the testing is.

dlex commented on src/v/config/bounded_property.h:
```diff
@@ -199,11 +269,9 @@ class bounded_property : public property<T> {
             // Take midpoint of min/max and align it.
-            guess = _bounds.min.value()
-                    + (_bounds.max.value() - _bounds.min.value()) / 2;
-            if (_bounds.align.has_value()) {
-                guess -= guess % _bounds.align.value();
-            }
```
align is done in `_bounds.clamp()` now. `numeric_integral_bounds::clamp()` does both clamp and align. Actual clamping is not needed in the call below since the value is the median, but the align part of the clamp is what it is there for.

Maybe `clamp` is not a good name for it and should be changed. In general, this function applies all the restriction laid by the Bounds class and produces a valid value as close to the argument as possible. Maybe `adjust`?

dotnwat replied:
I think it's all good. thanks for the clarification.

dlex commented:
> hoping too that you could characterize the testing that is in place for this feature. i see there are both UI and DT tests (yay), but it'd be nice to understand how aggressive/complete the testing is.

* a unit test is there to verify that the amounts consumed from memory semaphores are as expected under different conditions
* the `MemoryStressTest` is the test that reproduced the original issue, it verifies that the fix actually works
* for agressive testing, the rest of DT suite is relied upon, as there are numerous tests that actually consume, and some of them aggressively, and that was useful finding some issues already (which were close but unrelated)
* there is a good point how we can test the cross shard usage; unfortunately because of the kind of effects observed as the result, I have no idea if there can be a definitve test that would verify there is none. So far putting the entire DT into daily CI was what has worked best

dlex commented:
Force push:
* fixed: the semaphores were _still_ taken from the connection shard
* fixed: replica_selector instance used in fetch_from_ntp was from the connection shard
* fixed: `server_configuration` was only initialized in one shard in the test fixture

dotnwat commented on src/v/config/configuration.cc:
```diff
@@ -2074,7 +2074,15 @@ configuration::configuration()
+      "kafka_memory_share_for_fetch",
+      "The share of kafka subsystem memory that can be used for fetch read "
+      "buffers, as a fraction of kafka subsystem memory amount",
+      {.needs_restart = needs_restart::yes, .visibility = visibility::user},
```
we can definitely do this as a follow up ticket. just wondering if it would work alright because stuff like this feels like it could be an important property to change at runtime on occasion.

dotnwat commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -429,23 +566,27 @@ handle_shard_fetch(ss::shard_id shard, op_context& octx, shard_fetch fetch) {
       .invoke_on(
         shard,
         octx.ssg,
-        [foreign_read,
-         deadline = octx.deadline,
-         configs = std::move(fetch.requests),
-         &octx](cluster::partition_manager& mgr) mutable {
+        [foreign_read, configs = std::move(fetch.requests), &octx](
+          cluster::partition_manager& mgr) mutable {
+            // &octx is captured only to immediately use its accessors here so
+            // that there is a list of all objects accessed next to `invoke_on`.
+            // This is meant to help avoiding unintended cross shard access
             return fetch_ntps_in_parallel(
```
it looks like the semaphore reference captured is now correct, but IIUC the returned units will have captured a reference to that semaphore (so that the units' destructor can return the units to the semaphore), and won't that destructor run on the original shard?

dlex commented:
Force push: release semaphore units in the shard where they have been acquired.

dlex commented:
CI failures in https://buildkite.com/redpanda/redpanda/builds/30485 debug DT:
* #11044 (2 times)
* #11166 (3 times)
* #10497 (3 times)
* #8333
* #11151 (10 times)

vbotbuildovich commented:
/backport v23.1.x

vbotbuildovich commented:
/backport v22.3.x

vbotbuildovich commented:
Failed to run cherry-pick command. I executed the commands below:

```
git checkout -b backport-pr-10905-v22.3.x-170 remotes/upstream/v22.3.x
git cherry-pick -x 1db2edcd91fd7e692d069b1c059a03c353a8c689 ee5e069e3623dd655688df2d2db75c59a11b63f5 eb8a915a9497e6594ba712b0de58109d176c68bd d891efe7fc3dbc0ef27b17a53946336e17759296 58a9de0f0e32537f879404f3a4633fd5e94ed879 7b3860153a544aedd97a94a4d221ae034cf0439d d3d9276237a4c6f4ff3246c85ad3af40fc1e522e c1d77cd216d78e935586fac16274671329b70d97 2474ef67ed1e078cab1011e8835a47e71cede509 623e6136b35c4331a0a39aed1d9508edb8a7f5e0 950abc7d8ddaff656cfd035a061a999479b0a09e 06a38b97cfde550dcde20f6989ebc150750923e7
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/5195957497).

vbotbuildovich commented:
Failed to run cherry-pick command. I executed the commands below:

```
git checkout -b backport-pr-10905-v23.1.x-279 remotes/upstream/v23.1.x
git cherry-pick -x 1db2edcd91fd7e692d069b1c059a03c353a8c689 ee5e069e3623dd655688df2d2db75c59a11b63f5 eb8a915a9497e6594ba712b0de58109d176c68bd d891efe7fc3dbc0ef27b17a53946336e17759296 58a9de0f0e32537f879404f3a4633fd5e94ed879 7b3860153a544aedd97a94a4d221ae034cf0439d d3d9276237a4c6f4ff3246c85ad3af40fc1e522e c1d77cd216d78e935586fac16274671329b70d97 2474ef67ed1e078cab1011e8835a47e71cede509 623e6136b35c4331a0a39aed1d9508edb8a7f5e0 950abc7d8ddaff656cfd035a061a999479b0a09e 06a38b97cfde550dcde20f6989ebc150750923e7
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/5195957615).
